src/v/raft/mux_state_machine.h

```diff
@@ -289,6 +283,7 @@ ss::future<> mux_state_machine<T...>::apply(model::record_batch b) {
                 if (auto it = _promises.find(last_offset);
                     it != _promises.end()) {
                     it->second.set_value(ec);
+                    _promises.erase(it);
```

Comment:
How the double set was possible? Does it mean that apply may be called twice for the same offset?

Reply:
Doesn't it still possible? Now we don't remove a promise once it timed out so it's possible to double set it when apply finally catches up with it.

Another thing, I don't understand why the racing was possible in the first case, we check if a promise is available inside expiring_promise so it should be ok to double set it. Is `promise.available()` broken?

Reply:
Yep, expiring_promies (shared_promise) is [broken](https://github.com/vectorizedio/redpanda/commit/5827ee035804c5fd0cbff1fa3e1320c73115ea9e).

Reply:
I'll send a PR today fixing it too.

Reply:
Sent a [PR](https://github.com/vectorizedio/redpanda/pull/2471)